### DISCUSSION TITLE
Fallback en mode automatique ?

### DISCUSSION THREAD

**Discussion by tristanX2** [AmauriC/tarteaucitron.js#867] [Q&A]
Bonjour,

en mode automatique (version payante) comment utiliser la fonction fallback (pour un service déjà intégré comme googlemaps par exemple) ?

**Comment by AmauriC** [comment]
Le fallback n'est pas le même avec le mode auto, le fonctionnement est totalement différent.

Il y a peut-être une solution, quelle est la problématique?

**Reply by tristanX2** [reply]
j'utilise un service externe, qui lui utilise google maps (je n'utilise donc pas google maps directement sur mon site).

Le problème : si l'utilisateur a refusé le service googlemaps, le service externe ne fonctionne pas du tout et rien ne s'affiche.... Je voudrais donc dans ce cas, afficher un bouton du genre "Accepter Google Maps pour utiliser ce service"

**Reply by AmauriC** [reply]
Ok. Il n'y aura pas de solution toute prête mais vous pouvez vérifier en JS si le service est accepté : `if (tarteaucitron.state.googlemaps === true) { // retirer le message de fallback }`

Ou utiliser l'event:
```
document.addEventListener("googlemaps_loaded", function() {

   // google maps est dispo

});
```

**Reply by tristanX2** [reply]
merci Amauri pour ta réponse :-)

1) j'ai fait quelques tests, il semble qu'il faille plutôt utiliser "googlemaps_added" (au lieu de "googlemaps_loaded") ?
Tu peux confirmer ?
```
document.addEventListener('googlemaps_added', function() {
			var elem = document.getElementById('googlemaps_forbiden');
			if (elem) {
				if (tarteaucitron.state.googlemaps === true) {
					elem.style.display = 'none';
				} else {
					elem.style.display = 'block';
				}
			}
	});
```
Cela fonctionne bien, sauf dans un seul cas (bug ?) : si à chaud (clique sur l'icône "cookie" en bas à droite de l'écran), j'autorise "Google Maps" (ou tous les services),  l'évènement (googlemaps_added) n'est pas appelé :-( 
Dans le sens inverse (si je refuse "Google Maps" à chaud), cela fonctionne bien... 
Cela fonctionne aussi correctement, si je recharge la page...
Je peux éventuellement t'envoyer en MP (ou par email) l'adresse de la page si tu veux jeter un œil ?


2) comment afficher un bouton pour autoriser "Google Maps" ?
J'ai essayé ceci mais cela ne fonctionne pas :-(
`<button id=\"Eng50732edgooglemapsbutton\" class=\"tarteaucitronAllow\">Autoriser Google Maps</button>`

**Reply by AmauriC** [reply]
`_added` c'est quand tarteaucitron detecte google maps. La première partie est ok.
Ajoutes la même fonction avec `_loaded` et ça devrait corriger le problème de première page (la fonction sera rappelé via _loaded quand l'utilisateur accepte).

Pour la div, c'est mieux d'ajouter un onclick avec : `tarteaucitron.userInterface.respond(document.getElementById('googlemapsAllowed'), true);`

**Reply by tristanX2** [reply]
Pour le bouton, ok merci super ! 
Cela fonctionne :-)

Par contre, j'ai ajouté l'évènement googlemaps_loaded et même problème.
Si je passe à chaud (sur la même page) de googlemaps "non autorisé" à googlemaps "autorisé", les évènements googlemaps_loaded et googlemaps_added ne sont pas appelés :-( 
Si ensuite, j'actualise la page, cela fonctionne bien  :-( 
Si je passe à chaud de googlemaps  "autorisé" à googlemaps  "non autorisé" c'est bon également.

mon code complet :
```
<script>
	document.addEventListener('googlemaps_loaded', function() {
			console.log('googlemaps_loaded - state='+tarteaucitron.state.googlemaps);
			showHideServiceEnabled('googlemaps');
	});
	document.addEventListener('googlemaps_added', function() {
			console.log('googlemaps_added - state='+tarteaucitron.state.googlemaps);
			showHideServiceEnabled('googlemaps');
	});
	function showHideServiceEnabled(service) {
			var elem = document.getElementById(service+'_forbiden');
			if (elem) {
				if (tarteaucitron.state[service] === true) {
					elem.style.display = 'none';
				} else {
					elem.style.display = 'block';
				}
			}
	}
</script>

<p id="googlemaps_forbiden" style="display:none;">
<span class="tarteaucitron-magic-block">
<b>Google Maps est désactivé.</b><br>&nbsp;<br>
<br>&nbsp;<br>Pour activer ce service, merci d'autoriser Google Maps et d'accepter le dépôt, la lecture de cookies et l'utilisation de technologies de suivi nécessaires à son bon fonctionnement.
<button class="tarteaucitronAllow" onclick="tarteaucitron.userInterface.respond(document.getElementById('googlemapsAllowed'), true);document.getElementById('googlemaps_forbiden').style.display='none';">Autoriser Google Maps</button>
</span>
</p>
```

**Reply by AmauriC** [reply]
🧐 Possible d'avoir l'url de la page?

**Reply by tristanX2** [reply]
oui :-)  je viens de t'envoyer cela par email ...

**Reply by AmauriC** [reply]
Rien reçu : /
amauri@champeaux.fr

**Comment by AmauriC** [comment]
J'ai ajouté des events pour aider à régler ce problème https://github.com/AmauriC/tarteaucitron.js/commit/ab779fa14dfe0d6601e54cf12e605bc4ab5bca18
Ce sera dispo début de semaine sur la v pro ; )

**Reply by tristanX2** [reply]
Un exemple complet et fonctionnel (ici avec le service Google Maps).

Code javascript :
```
<script src=\"https://tarteaucitron.io/load.js?domain=www.domaine.com&uuid=xxxxxxxxxxxxxxxxx\"></script>
<script>
document.addEventListener('googlemaps_loaded', function() {
	showHideServiceEnabled('googlemaps');
});
document.addEventListener('googlemaps_added', function() {
	showHideServiceEnabled('googlemaps');
});
document.addEventListener('googlemaps_allowed', function() {
	tarteaucitron.state.googlemaps = true;
	showHideServiceEnabled('googlemaps');
});
document.addEventListener('googlemaps_disallowed', function() {
	tarteaucitron.state.googlemaps = false;
	showHideServiceEnabled('googlemaps');
});
function showHideServiceEnabled(service) {
        var elem = document.getElementById(service+'_forbiden');
	if (elem) {
		if (tarteaucitron.state[service] === true) {
			elem.style.display = 'none';
		} else {
			elem.style.display = 'block';
		}
	}
}	
</script>
```

Code HTML à placer dans la page :
```
<p id="googlemaps_forbiden" style="display: none; text-align: center;">
 <span class=" tarteaucitron-magic-block"><b>Google Maps est désactivé.</b><br>&nbsp;<br>
Ce service nécessite Google Maps pour son fonctionnement.
<br>&nbsp;<br>Merci d'autoriser Google Maps et d'accepter le dépôt, la lecture de cookies et l'utilisation de technologies de suivi nécessaires à son bon fonctionnement.
<button class="tarteaucitronAllow" onclick="tarteaucitron.userInterface.respond(document.getElementById('googlemapsAllowed'), true);document.getElementById('googlemaps_forbiden').style.display='none';">Autoriser Google Maps</button>
</span>
</p>
```

Merci Amauri pour le coup de main :-)

**Reply by AmauriC** [reply]
Merci à toi pour le retour, content qu'on ai trouvé une solution ; )
